### PR TITLE
Swift SIL and Optimizer: improvements, refactoring and bug fixes

### PR DESCRIPTION
**Pull request by eeckstein**
This PR contains a set of refactorings to simplify the code, improvements, bug fixes and new APIs.
For example, a rework of the PassContext.

For details, see the commit messages.

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: SwiftCompilerSources/Sources/Optimizer/DataStructures/Worklist.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// A utility for processing entities in a worklist.
+///
+/// A `Worklist` is basically a combination of a stack and a set.
+/// It can be used for typical worklist-processing algorithms.
+///
```
Comment (atrick) [inline SwiftCompilerSources/Sources/Optimizer/DataStructures/Worklist.swift:19]: @eeckstein In addition to the "standard" push/pop worklist I sometimes find it convenient to process a worklist in breadth-first order without popping, using an advancing iterator. That's very useful when you need a list of all the visited nodes as a result of processing the worklist.

**Review by atrick — COMMENTED** [review]
@eeckstein I quickly scanned these 20 commits. They all look like obviously good improvements to me.

File: SwiftCompilerSources/Sources/Optimizer/DataStructures/Worklist.swift (hunk 1 above)
Comment (eeckstein) [reply SwiftCompilerSources/Sources/Optimizer/DataStructures/Worklist.swift:19]: Yeah, we can add that when we need it

**Pull request merged**